### PR TITLE
Deprecating :opaque as an attribute, introducing it as a modifier

### LINKED ISSUE #3461: Deprecate using 'opaque' as an identifier
### Summary

We want 'opaque' to be a reserved word and to replace {:opaque}

### Background and Motivation

We want 'opaque' to be a reserved word and to replace {:opaque}

### Proposed Feature

First step is a warning about opaque being used as an identifier.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #3461

Still needs a news item and documentation and tests for where an opaque modifier may not be.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/AST/MemberDecls.cs [hunk 1]
```diff
@@ -14,6 +14,8 @@ public abstract class MemberDecl : Declaration {
       return HasStaticKeyword || (EnclosingClass is ClassDecl && ((ClassDecl)EnclosingClass).IsDefaultClass);
     }
   }
+
+  public bool IsOpaque;
```
Comment (RustanLeino) [inline Source/DafnyCore/AST/MemberDecls.cs:18]: Can this be `readonly`?

File: Source/DafnyCore/Dafny.atg [hunk 2]
```diff
@@ -253,7 +255,8 @@ public Parser(Scanner/*!*/ scanner, Errors/*!*/ errors, Include include, ModuleD
 }
 
 bool IsIdentifier(int kind) {
-  return kind == _ident || kind == _least || kind == _greatest || kind == _older;
+  if (kind == _opaque) errors.Deprecated(t, "opaque is deprecated as an identifier. It will soon become a reserved word. Use a different name.");
```
Comment (RustanLeino) [inline Source/DafnyCore/Dafny.atg:258]: The `IsIdentifier` method is not the place to output any messages. This line should be deleted. The deprecation message should only be output in the `Ident` grammar production (as you do below).

File: Test/git-issues/git-issue-2612B.dfy [hunk 3]
```diff
@@ -0,0 +1,8 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-2612B.dfy:1]: Can you make the `B` at the end of this filename lowercase? That's the style we've used before.

File: Test/git-issues/git-issue-2612B.dfy [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+// RUN: %diff "%s.expect" "%t"
+
+trait T {
+  opaque predicate True() { true }
+}
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-2612B.dfy:6]: You could easily add to this test file something that confirms that the `opaque` keyword is recognized. Something like

``` dafny
method TestOpaque(c: C) {
  assert c.True(); // error: True is opaque and hasn't been revealed
}

method TestRevealed(c: C) {
  reveal c.True(); // test that True can be revealed
  assert c.True(); // now, the assertion goes through
}
```

File: Test/git-issues/git-issue-370B.dfy [hunk 5]
```diff
@@ -0,0 +1,47 @@
+// RUN: %exits-with 4 %dafny /compile:0 "%s" > "%t"
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-370B.dfy:1]: Use lowercase `b` in filename.

**Review by RustanLeino — CHANGES_REQUESTED** [review]

File: Source/DafnyCore/AST/MemberDecls.cs (hunk 1 above)
Comment (davidcok) [reply Source/DafnyCore/AST/MemberDecls.cs:18]: Yes. Done.

File: Source/DafnyCore/Dafny.atg (hunk 2 above)
Comment (davidcok) [reply Source/DafnyCore/Dafny.atg:258]: Right - I intended to remove this one.

File: Test/git-issues/git-issue-2612B.dfy (hunk 3 above)
Comment (davidcok) [reply Test/git-issues/git-issue-2612B.dfy:1]: Done

File: Test/git-issues/git-issue-370B.dfy (hunk 5 above)
Comment (davidcok) [reply Test/git-issues/git-issue-370B.dfy:1]: Done

File: Test/git-issues/git-issue-2612B.dfy (hunk 4 above)
Comment (davidcok) [reply Test/git-issues/git-issue-2612B.dfy:6]: Added.

File: Source/DafnyCore/AST/MemberDecls.cs [hunk 6]
```diff
@@ -15,7 +15,10 @@ public abstract class MemberDecl : Declaration {
     }
   }
 
-  public readonly bool IsOpaque;
+  public virtual bool IsOpaque {
+    get { return false; }
+  }
```
Comment (RustanLeino) [inline Source/DafnyCore/AST/MemberDecls.cs:20]: C# syntax allows you to write this as

```suggestion
  public virtual bool IsOpaque => false;
```

File: Test/git-issues/git-issue-3461.dfy [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+opaque function f(): int { 42 } // OK
+opaque predicate p() { true } // OK
+
+opaque class A {} // NO
+opaque datatype D = D // NO
+opaque newtype N = int // NO
+opaque type P = i | i >= 0 // NO
+
+method z() {
+  opaque var j: int; // NO
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-3461.dfy:17]:
```suggestion
  opaque var j: int // NO
```

File: docs/DafnyRef/Types.md [hunk 8]
```diff
@@ -2083,7 +2083,7 @@ which value is fixed after initialization.
 The declaration must either have a type or an initializing expression (or both).
 If the type is omitted, it is inferred from the initializing expression.
 
-* A const declaration may include the `ghost` and `static` modifiers, but no
+* A const declaration may include the `ghost`, `static` and `opaque` modifiers, but no
```
Comment (RustanLeino) [inline docs/DafnyRef/Types.md:2086]:
```suggestion
* A const declaration may include the `ghost`, `static`, and `opaque` modifiers, but no
```

File: docs/DafnyRef/Types.md [hunk 9]
```diff
@@ -2092,6 +2092,8 @@ that may contain members (class, trait, datatype, newtype).
 * If the declaration has an initializing expression that is a ghost
 expression, then the ghost-ness of the declaration is inferred; the `ghost`
 modifier may be omitted.
+* If the declaration includes the `opaque` modifier, then uses of the declared
+variable know its name and type but not its value .
```
Comment (RustanLeino) [inline docs/DafnyRef/Types.md:2096]:
```suggestion
variable know its name and type but not its value.
```

File: docs/DafnyRef/Types.md [hunk 10]
```diff
@@ -2696,6 +2698,11 @@ default. To make it ghost, replace the keyword `function` with the two keywords
 (See the [/functionSyntax option](#sec-function-syntax) for a description 
 of the migration path for this change in behavior.}
 
+Functions (including predicates, function-by-methods, two-state functions and extreme predicates) may be
```
Comment (RustanLeino) [inline docs/DafnyRef/Types.md:2701]:
```suggestion
Functions (including predicates, function-by-methods, two-state functions, and extreme predicates) may be 
```

File: docs/DafnyRef/Types.md (hunk 9 above)
Comment (RustanLeino) [reply docs/DafnyRef/Types.md:2096]: Also, mention (here?) that an opaque constant can be revealed.

File: Source/DafnyCore/AST/Function.cs [hunk 11]
```diff
@@ -33,6 +33,12 @@ public class Function : MemberDecl, TypeParameter.ParentType, ICallable {
     }
   }
 
+  private bool IsOpaque_;
```
[the remark this answers is not in the document]
Comment (davidcok) [reply Source/DafnyCore/AST/Function.cs:36]: Fine. Done - though that is not the style in the rest of this class!

File: docs/DafnyRef/Types.md (hunk 8 above)
Comment (davidcok) [reply docs/DafnyRef/Types.md]: Oxford, here we come

**Review by RustanLeino — APPROVED** [review]

**Comment by MikaelMayer** [comment]
Wooohooo exciting !

**Pull request merged**